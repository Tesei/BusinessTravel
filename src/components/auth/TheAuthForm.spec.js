import { describe, it, expect, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import flushPromises from 'flush-promises'
import waitForExpect from 'wait-for-expect'
import { useUserStore } from '@/store/user/index.js'
import component from './TheAuthForm.vue'
import { useRouter } from 'vue-router'

vi.mock('vue-router', () => ({
    useRoute: vi.fn(),
    useRouter: vi.fn(() => ({
        push: () => {},
    })),
}))

describe('TheAuthForm component', () => {
    // ? stubActions - будeт ли компонент подключён (true/false) к стору
    const buildWrapper = (stubActions = false) => {
        return mount(component, {
            global: {
                stubs: ['router-link', 'router-view'], // Stubs for router-link and router-view in case they're rendered in your template
                plugins: [createTestingPinia({ stubActions })],
            },
        })
    }

    it('mounts without error', () => {
        const wrapper = buildWrapper()
        expect(wrapper.exists).toBeTruthy()
        expect(wrapper.html()).toMatchSnapshot()
    })
    it('проверка заблокированной кнопки при отсутствии данных ввода', async () => {
        const wrapper = buildWrapper()
        const loginInput = wrapper.find('input[type="text"]')
        const passwordInput = wrapper.find('input[type="password"]')

        await loginInput.setValue('')
        await passwordInput.setValue('')
        expect(wrapper.find('button').attributes().disabled).toBe('')
    })
    it('наличие ссылки на /restore-password', async () => {
        const wrapper = buildWrapper()        
        expect(wrapper.find('.auth__forgot-password').attributes('href')).toBe('/restore-password')

    })

    // Пример тестирования с замоканым роутом
    it('проверка перехода на WelcomePage', async () => {
        //   useRoute.mockImplementationOnce(() => ({
        //     params: {
        //       id: 1
        //     }
        //   }))
        const push = vi.fn()
        useRouter.mockImplementationOnce(() => ({
            push,
        }))
        const wrapper = buildWrapper(true)
        const store = useUserStore() // ? должно стоять после wrapper

        store.userAuthorized = true
        store.fetchLogIn.mockResolvedValue({ accessToken: '23112312' })
        store.uploadProfile.mockResolvedValue({ permissions: ['1', '2'] })

        const loginInput = wrapper.find('input[type="text"]')
        const passwordInput = wrapper.find('input[type="password"]')

        await loginInput.setValue('')
        await passwordInput.setValue('')
        await loginInput.setValue('testad1@mail.ru')
        await passwordInput.setValue('1342Df!352353532532')
        await wrapper.get('form').trigger('submit') // working !

        await flushPromises()
        await waitForExpect(() => {
            // ? Проверка вызова функций:
            expect(store.fetchLogIn).toHaveBeenCalledTimes(1)
            expect(store.fetchLogIn).toHaveBeenCalledWith({
                login: 'testad1@mail.ru',
                password: '1342Df!352353532532',
            })
            expect(store.uploadProfile).toHaveBeenCalledTimes(1)
            
            // ? Проверка перехода по верному роуту:
            expect(push).toHaveBeenCalledTimes(1)
            expect(push).toHaveBeenCalledWith({ name: 'WelcomePage' })
        })
    })

    it('проверка перехода на /incorrect-profile', async () => {
        const push = vi.fn()
        useRouter.mockImplementationOnce(() => ({
            push,
        }))
        const wrapper = buildWrapper(true)
        const store = useUserStore() // ? должно стоять после wrapper
        store.userAuthorized = true
        store.fetchLogIn.mockResolvedValue({ accessToken: '23112312' })
        store.uploadProfile.mockResolvedValue({ permissions: [] })

        const loginInput = wrapper.find('input[type="text"]')
        const passwordInput = wrapper.find('input[type="password"]')

        await loginInput.setValue('')
        await passwordInput.setValue('')
        await loginInput.setValue('testad1@mail.ru')
        await passwordInput.setValue('1342Df!352353532532')
        await wrapper.get('form').trigger('submit') // working !

        await flushPromises()
        await waitForExpect(() => {
            // ? Проверка вызова функций:
            expect(store.fetchLogIn).toHaveBeenCalledTimes(1)
            expect(store.fetchLogIn).toHaveBeenCalledWith({
                login: 'testad1@mail.ru',
                password: '1342Df!352353532532',
            })
            expect(store.uploadProfile).toHaveBeenCalledTimes(1)
            
            // ? Проверка перехода по верному роуту:
            expect(push).toHaveBeenCalledTimes(1)
            expect(push).toHaveBeenCalledWith('/incorrect-profile')
        })
    })
})
