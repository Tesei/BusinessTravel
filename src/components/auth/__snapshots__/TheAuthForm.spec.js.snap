// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheAuthForm component > mounts without error 1`] = `
"<div data-v-5f48b7ce="">
  <div data-v-400d8eb7="" data-v-5f48b7ce="" class="auth">
    <div data-v-400d8eb7="" class="ui-login__wrapper">
      <div data-v-400d8eb7="" class="bd radius-16 p-24 bg-white bd-gray-9 ui-login__content">
        <div data-v-400d8eb7="" class="ui-login__header ui-login__content_header">
          <h3 data-v-5f48b7ce="" class="Heading_3 block mb-24 auth__title">Авторизация</h3>
        </div>
        <div data-v-400d8eb7="" class="ui-login__body">
          <form data-v-5f48b7ce="" class="auth__inputs-block">
            <div data-v-2b79102b="" data-v-5f48b7ce="" class="input input-default input-container body_S mb-24 auth__input-login">
              <div data-v-2b79102b="" class="input__wrapper">
                <div data-v-2b79102b="" class="input__wrapper"><input data-v-2b79102b="" class="body_S input__field" type="text"><label data-v-2b79102b="" class="body_S input__placeholder required">Логин</label></div>
                <!--v-if-->
              </div>
            </div>
            <div data-v-b93d3001="" data-v-5f48b7ce="" class="input input-default body_S mb-24 auth__input-password">
              <div data-v-b93d3001="" class="input__wrapper"><input data-v-b93d3001="" class="body_S input__field" type="password"><label data-v-b93d3001="" class="body_S input__placeholder required">Пароль</label><svg data-v-b93d3001="" class="icon-sprite input__svg input__svg-eye">
                  <use xlink:href="/assets/icons/sprite.svg#common--eye-hide"></use>
                </svg></div>
              <!--v-if-->
            </div><a data-v-5f48b7ce="" href="/restore-password" class="body_S mb-24 c-gray auth__forgot-password">Забыли пароль? </a><button data-v-170f5cbe="" data-v-5f48b7ce="" class="button font-button_M jc-c font-button_M button button_blue jc-c auth__btn" disabled="" type="submit" iconname="common--delete"> Войти </button>
          </form>
        </div>
        <!--v-if-->
        <div data-v-400d8eb7="" class="ui-login__footer"></div>
      </div>
    </div>
  </div>
</div>"
`;
